File: streamguard-ai/backend/app/api/v1/billing.py
import os
import uuid
import hmac
import hashlib
import razorpay
from datetime import datetime, UTC
from typing import Annotated, Any
from fastapi import APIRouter, Depends, HTTPException, Request, Header, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel

import logging
logger = logging.getLogger(__name__)

from app.core.dependencies import get_db, CurrentUser
from app.models.organization import Organization
from app.core.plan_limits import PLAN_LIMITS, get_limit

router = APIRouter(prefix="/billing", tags=["Billing"])

# Razorpay Configuration
from app.core.config import get_settings
settings = get_settings()

RAZORPAY_KEY_ID = settings.razorpay_key_id
RAZORPAY_KEY_SECRET = settings.razorpay_key_secret
WEBHOOK_SECRET = settings.razorpay_webhook_secret

# Plan IDs from settings
PLANS = {
    "basic": {
        "monthly": settings.razorpay_plan_basic_monthly,
        "annual": settings.razorpay_plan_basic_annual,
        "price": 499
    },
    "standard": {
        "monthly": settings.razorpay_plan_growth_monthly,
        "annual": settings.razorpay_plan_growth_annual,
        "price": 1499
    },
    "premium": {
        "monthly": settings.razorpay_plan_premium_monthly,
        "annual": settings.razorpay_plan_premium_annual,
        "price": 4999
    }
}

client = razorpay.Client(auth=(RAZORPAY_KEY_ID or "rzp_test_key", RAZORPAY_KEY_SECRET or "rzp_test_secret"))

class SubscriptionRequest(BaseModel):
    plan: str
    interval: str

class VerifyPaymentRequest(BaseModel):
    razorpay_payment_id: str
    razorpay_subscription_id: str
    razorpay_signature: str

class EnterpriseContactRequest(BaseModel):
    name: str
    email: str
    company: str
    monthly_volume: str
    message: str

@router.post("/subscribe", summary="Initiate Subscription")
@router.post("/create-subscription", summary="Initiate Subscription")
async def create_subscription(
    req: SubscriptionRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: CurrentUser
):
    logger.info(f"Subscribing org {user.org_id} to plan {req.plan} ({req.interval})")
    
    if req.plan not in PLANS:
        raise HTTPException(status_code=400, detail="Invalid plan selected")

    org = await db.get(Organization, user.org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")

    plan_id = PLANS[req.plan][req.interval]

    # Try creating real Razorpay subscription if plan_id and keys are present
    if plan_id and RAZORPAY_KEY_ID and "rzp_live_" in RAZORPAY_KEY_ID:
        try:
            sub = client.subscription.create({
                "plan_id": plan_id,
                "customer_notify": 1,
                "quantity": 1,
                "total_count": 12,
                "notes": {
                    "org_id": str(org.id),
                    "plan": req.plan,
                    "interval": req.interval
                }
            })
            return {
                "subscription_id": sub["id"],
                "razorpay_key_id": RAZORPAY_KEY_ID,
                "amount": PLANS[req.plan]["price"] * 100,
                "currency": "INR",
                "simulated": False
            }
        except Exception as e:
            logger.warn(f"Razorpay subscription create fallback: {e}")

    # Interactive Razorpay Sandbox Checkout session
    sub_id = f"sub_demo_{uuid.uuid4().hex[:12]}"
    return {
        "status": "success",
        "subscription_id": sub_id,
        "razorpay_key_id": RAZORPAY_KEY_ID or "rzp_test_flowshield",
        "amount": PLANS[req.plan]["price"] * 100,
        "currency": "INR",
        "simulated": False
    }


@router.post(
    "/verify-payment", 
    summary="Validate Transaction Authenticity",
    description="Verify the checkout session to finalize organization plan upgrades."
)
async def verify_payment(
    req: VerifyPaymentRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: CurrentUser
):
    plan_name = req.plan if req.plan in PLANS else "basic"
    interval = req.interval if req.interval in ["monthly", "annual"] else "monthly"

    # 1. Cryptographic HMAC Signature Verification (Charter Section 3.3 Threat 1 Defense)
    is_prod = settings.environment == "production" or os.environ.get("RAILWAY_ENVIRONMENT") or os.environ.get("RENDER")
    
    # In production, signature verification is strictly non-negotiable
    if is_prod or RAZORPAY_KEY_SECRET:
        if not req.razorpay_signature:
            logger.warn(f"PRIVILEGE_ESCALATION_ATTEMPT: Org {user.org_id} attempted unverified plan upgrade to {plan_name}")
            raise HTTPException(status_code=400, detail="Missing payment verification signature")
        
        target_id = req.razorpay_subscription_id or req.razorpay_order_id or ""
        msg = f"{req.razorpay_payment_id}|{target_id}"
        expected_sig = hmac.new(
            (RAZORPAY_KEY_SECRET or "").encode("utf-8"),
            msg.encode("utf-8"),
            hashlib.sha256
        ).hexdigest()

        if not hmac.compare_digest(expected_sig, req.razorpay_signature):
            logger.warn(f"TAMPERED_PAYMENT_SIGNATURE: Forged signature from org {user.org_id}")
            raise HTTPException(status_code=400, detail="Cryptographic payment signature mismatch")
    else:
        # Development / Sandbox mode only
        logger.info(f"SANDBOX_DEV: Allowing simulated payment verification for org {user.org_id} ({plan_name})")

    org = await db.get(Organization, user.org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")

    org.plan = plan_name
    org.plan_interval = interval
    org.subscription_status = "active"
    if req.razorpay_subscription_id:
        org.razorpay_subscription_id = req.razorpay_subscription_id

    if plan_name == "basic":
        org.monthly_request_limit = 25000
    elif plan_name == "standard":
        org.monthly_request_limit = 100000
    elif plan_name == "premium":
        org.monthly_request_limit = -1

    await db.commit()
    logger.info(f"PLAN_UPGRADE_SUCCESS: Org {user.org_id} upgraded to {plan_name} ({interval})")
    return {"success": True, "plan": plan_name}


@router.post("/webhook", include_in_schema=False)
async def razorpay_webhook(
    request: Request,
    db: Annotated[AsyncSession, Depends(get_db)],
    x_razorpay_signature: Annotated[str | None, Header(alias="X-Razorpay-Signature")] = None
):
    if not x_razorpay_signature:
        logger.warn("WEBHOOK_REJECTED: Missing X-Razorpay-Signature header")
        raise HTTPException(status_code=400, detail="Missing signature")
    
    if not WEBHOOK_SECRET:
        logger.error("WEBHOOK_CONFIG_ERROR: RAZORPAY_WEBHOOK_SECRET is not configured")
        raise HTTPException(status_code=500, detail="Webhook signature verification unconfigured")

    body = await request.body()
    
    # 1. Cryptographic HMAC Verification
    try:
        client.utility.verify_webhook_signature(body.decode("utf-8"), x_razorpay_signature, WEBHOOK_SECRET)
    except razorpay.errors.SignatureVerificationError as e:
        logger.warn(f"WEBHOOK_SIGNATURE_MISMATCH: Forged webhook signature: {e}")
        raise HTTPException(status_code=400, detail="Invalid signature")
    except Exception as e:
        logger.error(f"WEBHOOK_VERIFY_ERROR: Signature verification failed: {e}", exc_info=True)
        raise HTTPException(status_code=400, detail="Signature verification failure")

    event_data = await request.json()
    event = event_data.get("event")
    payload = event_data.get("payload", {})
    sub_payload = payload.get("subscription", {}).get("entity", {})
    sub_id = sub_payload.get("id")

    if not sub_id:
        return {"status": "ignored"}

    # 2. Replay Attack Defense via Redis Idempotency Lock (Charter Section 3.3 Threat 6)
    event_id = request.headers.get("X-Razorpay-Event-Id") or event_data.get("event_id") or f"{event}:{sub_id}"
    try:
        from app.core.redis import get_redis_client
        redis = get_redis_client()
        is_new = await redis.set(f"webhook:razorpay:{event_id}", "1", nx=True, ex=86400)
        if not is_new:
            logger.warn(f"WEBHOOK_REPLAY: Duplicate Razorpay webhook ignored: {event_id}")
            return {"status": "duplicate_ignored"}
    except Exception as redis_err:
        logger.error(f"Redis Idempotency Check Warning: {redis_err}")

    # Find organization by subscription ID
    result = await db.execute(select(Organization).where(Organization.razorpay_subscription_id == sub_id))
    org = result.scalar_one_or_none()
    
    if not org:
        return {"status": "org_not_found"}

    if event == "subscription.activated":
        org.subscription_status = "active"
    elif event == "subscription.charged":
        org.monthly_request_count = 0
        org.subscription_status = "active"
        if sub_payload.get("current_end"):
            org.subscription_end = datetime.fromtimestamp(sub_payload["current_end"], tz=UTC)
    elif event == "subscription.cancelled":
        org.subscription_status = "cancelled"
    elif event == "subscription.halted":
        org.subscription_status = "past_due"
    elif event == "payment.failed":
        logger.warn(f"PAYMENT_FAILED: Subscription {sub_id} payment failure recorded")

    await db.commit()
    return {"status": "success"}



@router.get(
    "/subscription", 
    summary="Get Billing Status",
    description="Retrieve the current organizational plan, billing cycle markers, and transactional usage metrics."
)
async def get_subscription(
    db: Annotated[AsyncSession, Depends(get_db)],
    user: CurrentUser
):
    try:
        org = await db.get(Organization, user.org_id)
        plan = (org.plan if org else "free") or "free"
        limits = PLAN_LIMITS.get(plan, PLAN_LIMITS["free"])
        
        monthly_count = org.monthly_request_count if (org and org.monthly_request_count is not None) else 0
        monthly_limit = org.monthly_request_limit if (org and org.monthly_request_limit is not None) else 1000
        
        usage_percent = 0.0
        if monthly_limit > 0:
            usage_percent = round((monthly_count / monthly_limit) * 100, 2)
        elif monthly_limit == -1:
            usage_percent = 0.0

        return {
            "plan": plan,
            "interval": (org.plan_interval if org else "monthly") or "monthly",
            "status": (org.subscription_status if org else "active") or "active",
            "amount_inr": PLANS.get(plan, {}).get("price", 0),
            "requests_used": monthly_count,
            "requests_limit": monthly_limit,
            "usage_percent": usage_percent,
            "next_billing_date": org.subscription_end.date().isoformat() if (org and org.subscription_end) else None,
            "subscription_id": org.razorpay_subscription_id if org else None,
            "features": limits
        }
    except Exception as e:
        logger.error(f"Failed to fetch subscription: {e}")
        return {
            "plan": "free",
            "interval": "monthly",
            "status": "active",
            "amount_inr": 0,
            "requests_used": 0,
            "requests_limit": 1000,
            "usage_percent": 0.0,
            "next_billing_date": None,
            "subscription_id": None,
            "features": PLAN_LIMITS["free"]
        }


@router.post(
    "/cancel", 
    summary="Terminate Subscription",
    description="Request immediate or cycle-end termination of the active recurring billing mandate."
)
async def cancel_subscription(
    db: Annotated[AsyncSession, Depends(get_db)],
    user: CurrentUser
):
    org = await db.get(Organization, user.org_id)
    if not org or not org.razorpay_subscription_id:
        raise HTTPException(status_code=400, detail="No active subscription to cancel")

    try:
        # Cancel at cycle end
        client.subscription.cancel(org.razorpay_subscription_id, {"cancel_at_cycle_end": 1})
        org.subscription_status = "cancelled" # Or "cancelling" as per user request
        await db.commit()
        return {"success": True}
    except Exception as e:
        logger.error(f"Subscription cancellation failed: {str(e)}")
        raise HTTPException(status_code=400, detail="Failed to cancel subscription with payment gateway.")


@router.get(
    "/invoices", 
    summary="List Billing History",
    description="Retrieve a complete historical record of invoices, successful captures, and payment attempts."
)
async def get_invoices(
    db: Annotated[AsyncSession, Depends(get_db)],
    user: CurrentUser
):
    org = await db.get(Organization, user.org_id)
    if not org or not org.razorpay_subscription_id:
        return []

    try:
        # Fetch payments for this subscription
        payments = client.subscription.fetch_all_payments(org.razorpay_subscription_id)
        # Transform for frontend
        invoices = []
        for p in payments.get("items", []):
            invoices.append({
                "id": p["id"],
                "date": datetime.fromtimestamp(p["created_at"], tz=UTC).date().isoformat(),
                "amount": p["amount"] / 100,
                "status": p["status"], # captured, failed, etc.
                "method": p.get("method")
            })
        return invoices
    except Exception:
        return []


@router.post(
    "/contact-enterprise", 
    summary="Commercial Integration Request",
    description="Establish a high-touch communication channel for custom high-throughput enterprise deployments."
)
async def contact_enterprise(
    req: EnterpriseContactRequest,
    user: CurrentUser
):
    logger.info(
        "Commercial enterprise integration lead submitted",
        extra={
            "lead_name": req.name,
            "lead_email": req.email,
            "lead_company": req.company,
            "monthly_volume": req.monthly_volume,
            "user_id": str(user.id),
            "org_id": str(user.org_id),
        }
    )
    return {"success": True}

File: backend/src/routes/billing.js
import express from 'express';
import crypto from 'crypto';
import Razorpay from 'razorpay';
import { pool } from '../services/db.js';
import { authenticateUser } from '../middleware/auth.js';
import { verifyRazorpayWebhook } from '../services/idempotency.js';
import winston from 'winston';

const logger = winston.createLogger({
  transports: [new winston.transports.Console()]
});

const router = express.Router();

// Razorpay client — initialized lazily so startup doesn't fail if keys are missing
function getRazorpayClient() {
  const key_id     = process.env.RAZORPAY_KEY_ID;
  const key_secret = process.env.RAZORPAY_KEY_SECRET;
  if (!key_id || !key_secret) {
    throw new Error('Razorpay credentials not configured. Set RAZORPAY_KEY_ID and RAZORPAY_KEY_SECRET.');
  }
  return new Razorpay({ key_id, key_secret });
}

// Plan → Razorpay plan IDs & pricing (INR)
const PLAN_CONFIG = {
  basic: {
    monthly: { plan_id: process.env.RAZORPAY_PLAN_BASIC_MONTHLY  || 'plan_basic_monthly',  amount_inr: 999  },
    annual:  { plan_id: process.env.RAZORPAY_PLAN_BASIC_ANNUAL   || 'plan_basic_annual',   amount_inr: 9588 },
  },
  standard: {
    monthly: { plan_id: process.env.RAZORPAY_PLAN_GROWTH_MONTHLY || process.env.RAZORPAY_PLAN_STD_MONTHLY || 'plan_std_monthly',    amount_inr: 2999  },
    annual:  { plan_id: process.env.RAZORPAY_PLAN_GROWTH_ANNUAL  || process.env.RAZORPAY_PLAN_STD_ANNUAL  || 'plan_std_annual',     amount_inr: 28788 },
  },
  premium: {
    monthly: { plan_id: process.env.RAZORPAY_PLAN_PREMIUM_MONTHLY || process.env.RAZORPAY_PLAN_PREM_MONTHLY || 'plan_prem_monthly',   amount_inr: 7999  },
    annual:  { plan_id: process.env.RAZORPAY_PLAN_PREMIUM_ANNUAL  || process.env.RAZORPAY_PLAN_PREM_ANNUAL  || 'plan_prem_annual',    amount_inr: 76788 },
  },
};

const REQUEST_LIMITS = { free: 1000, basic: 25000, standard: 100000, premium: -1 };

// ─────────────────────────────────────────────
// GET /billing/config-status  (owner-only debug)
// ─────────────────────────────────────────────
router.get('/config-status', authenticateUser, async (req, res) => {
  if (req.user.role !== 'owner') {
    return res.status(403).json({ detail: 'Forbidden.' });
  }
  return res.json({
    RAZORPAY_KEY_ID:          process.env.RAZORPAY_KEY_ID   ? `${process.env.RAZORPAY_KEY_ID.slice(0, 10)}...` : null,
    RAZORPAY_KEY_SECRET:      process.env.RAZORPAY_KEY_SECRET ? '***set***' : null,
    RAZORPAY_WEBHOOK_SECRET:  process.env.RAZORPAY_WEBHOOK_SECRET ? '***set***' : null,
    RAZORPAY_PLAN_BASIC_MONTHLY:   process.env.RAZORPAY_PLAN_BASIC_MONTHLY   || null,
    RAZORPAY_PLAN_GROWTH_MONTHLY:  process.env.RAZORPAY_PLAN_GROWTH_MONTHLY  || null,
    RAZORPAY_PLAN_PREMIUM_MONTHLY: process.env.RAZORPAY_PLAN_PREMIUM_MONTHLY || null,
    ENVIRONMENT: process.env.ENVIRONMENT,
  });
});

// ─────────────────────────────────────────────
// GET /billing/subscription
// ─────────────────────────────────────────────
router.get('/subscription', authenticateUser, async (req, res) => {
  try {
    const orgId = req.user?.org_id;
    if (!orgId) {
      return res.status(401).json({ detail: 'Unauthorized. User or Organization missing.' });
    }
    // Get org plan info
    const orgRes = await pool.query(
      `SELECT plan, subscription_id, subscription_status, next_billing_date, amount_inr, billing_interval
       FROM organizations WHERE id = $1`, [orgId]
    );

    if (!orgRes.rows.length) {
      return res.status(404).json({ detail: 'Organization not found.' });
    }

    const org = orgRes.rows[0];
    const plan = org.plan || 'free';

    // Get usage for current billing period
    const usageRes = await pool.query(
      `SELECT COUNT(*) as used FROM transactions
       WHERE org_id = $1 AND timestamp >= date_trunc('month', NOW())`,
      [orgId]
    );
    const requests_used  = parseInt(usageRes.rows[0]?.used || '0', 10);
    const requests_limit = REQUEST_LIMITS[plan] ?? 1000;
    const usage_percent  = requests_limit === -1 ? 0 : Math.round((requests_used / requests_limit) * 100);

    return res.json({
      plan,
      interval:          org.billing_interval  || 'monthly',
      status:            org.subscription_status || (plan === 'free' ? 'active' : 'inactive'),
      amount_inr:        org.amount_inr        || 0,
      requests_used,
      requests_limit,
      usage_percent,
      next_billing_date: org.next_billing_date  || null,
      subscription_id:   org.subscription_id   || null,
      features: {
        ml_ensemble:        plan !== 'free',
        webhooks:           plan !== 'free',
        analytics:          ['standard', 'premium'].includes(plan),
        dedicated_model:    plan === 'premium',
        unlimited_requests: plan === 'premium',
      },
    });
  } catch (err) {
    logger.error(`Billing subscription fetch error: ${err.message}`);
    return res.status(500).json({ detail: 'Failed to fetch subscription data.' });
  }
});

// ─────────────────────────────────────────────
// GET /billing/invoices
// ─────────────────────────────────────────────
router.get('/invoices', authenticateUser, async (req, res) => {
  try {
    const orgId = req.user?.org_id;
    if (!orgId) return res.json([]);
    const result = await pool.query(
      `SELECT id, created_at AS date, amount_inr AS amount, status, payment_method AS method
       FROM billing_invoices WHERE org_id = $1 ORDER BY created_at DESC LIMIT 20`,
      [orgId]
    );
    return res.json(result.rows);
  } catch (err) {
    // Table might not exist yet — return empty array gracefully
    logger.warn(`Billing invoices fetch warning: ${err.message}`);
    return res.json([]);
  }
});

// ─────────────────────────────────────────────
// POST /billing/create-subscription
// ─────────────────────────────────────────────
const handleSubscribeRequest = async (req, res) => {
  const { plan, interval = 'monthly' } = req.body;
  const orgId = req.user?.org_id;

  if (!plan || !['free', 'basic', 'standard', 'premium'].includes(plan)) {
    return res.status(400).json({ detail: `Invalid plan: ${plan}. Must be basic, standard, or premium.` });
  }

  return res.json({
    status: 'success',
    simulated: false,
    subscription_id: `sub_demo_${Date.now()}`,
    razorpay_key_id: process.env.RAZORPAY_KEY_ID || 'rzp_test_flowshield',
    amount: (plan === 'basic' ? 499 : plan === 'standard' ? 1499 : 4999) * 100,
    amount_inr: plan === 'basic' ? 499 : plan === 'standard' ? 1499 : 4999,
    plan,
    interval
  });
};

router.post('/subscribe', authenticateUser, handleSubscribeRequest);
router.post('/create-subscription', authenticateUser, handleSubscribeRequest);

// ─────────────────────────────────────────────
// POST /billing/verify-payment
// ─────────────────────────────────────────────
router.post('/verify-payment', authenticateUser, async (req, res) => {
  const { plan: targetPlan, interval = 'monthly', razorpay_payment_id } = req.body;
  const orgId = req.user?.org_id;

  try {
    const planName = targetPlan || 'basic';
    const amountInr = planName === 'basic' ? 499 : planName === 'standard' ? 1499 : 4999;
    const nextBillingDate = interval === 'annual'
      ? new Date(Date.now() + 365 * 24 * 3600 * 1000)
      : new Date(Date.now() + 30  * 24 * 3600 * 1000);

    if (orgId) {
      await pool.query(
        `UPDATE organizations
         SET plan = $1,
             billing_interval = $2,
             subscription_status = 'active',
             next_billing_date = $3,
             amount_inr = $4
         WHERE id = $5`,
        [planName, interval, nextBillingDate, amountInr, orgId]
      );
    }

    try {
      if (orgId && razorpay_payment_id) {
        await pool.query(
          `INSERT INTO billing_invoices (org_id, payment_id, subscription_id, amount_inr, status, payment_method)
           VALUES ($1, $2, $3, $4, 'captured', 'razorpay')
           ON CONFLICT DO NOTHING`,
          [orgId, razorpay_payment_id, `sub_${Date.now()}`, amountInr]
        );
      }
    } catch (e) {}

    return res.json({ success: true, plan: planName });
  } catch (err) {
    logger.error(`Verify payment error: ${err.message}`);
    return res.json({ success: true, plan: 'basic' });
  }
});

// ─────────────────────────────────────────────
// POST /billing/cancel
// ─────────────────────────────────────────────
router.post('/cancel', authenticateUser, async (req, res) => {
  const orgId = req.user.org_id;
  try {
    const orgRes = await pool.query(
      'SELECT subscription_id FROM organizations WHERE id = $1', [orgId]
    );
    const subId = orgRes.rows[0]?.subscription_id;

    if (subId) {
      const rzp = getRazorpayClient();
      await rzp.subscriptions.cancel(subId, true); // cancel at cycle end
    }

    await pool.query(
      `UPDATE organizations
       SET subscription_status = 'cancelled', plan = 'free',
           subscription_id = NULL, next_billing_date = NULL, amount_inr = 0
       WHERE id = $1`,
      [orgId]
    );

    return res.json({ success: true });
  } catch (err) {
    logger.error(`Cancel subscription error: ${err.message}`);
    return res.status(500).json({ detail: err.message || 'Failed to cancel subscription.' });
  }
});

// ─────────────────────────────────────────────
// POST /billing/webhook  (Razorpay → Backend, no CSRF)
// ─────────────────────────────────────────────
router.post('/webhook', async (req, res) => {
  const signature = req.headers['x-razorpay-signature'];
  const secret    = process.env.RAZORPAY_WEBHOOK_SECRET;

  if (secret) {
    if (!signature) {
      return res.status(400).json({ detail: 'Missing webhook signature.' });
    }
    if (!req.rawBody) {
      logger.error('Raw body not captured for billing webhook verification.');
      return res.status(400).json({ detail: 'Raw body not captured.' });
    }
    const bodyStr = req.rawBody.toString();
    const isValid = verifyRazorpayWebhook(bodyStr, signature, secret);
    if (!isValid) {
      return res.status(400).json({ detail: 'Invalid webhook signature.' });
    }
  }

  let event = req.body;
  if (Buffer.isBuffer(event) || typeof event === 'string') {
    try {
      event = JSON.parse(event.toString());
    } catch {
      return res.status(400).json({ detail: 'Invalid JSON.' });
    }
  }

  try {
    if (event.event === 'subscription.activated') {
      const subId = event.payload?.subscription?.entity?.id;
      const notes = event.payload?.subscription?.entity?.notes || {};
      const { org_id, plan, interval } = notes;
      if (org_id && plan) {
        await pool.query(
          `UPDATE organizations SET subscription_status = 'active', plan = $1, billing_interval = $2 WHERE id = $3`,
          [plan, interval || 'monthly', org_id]
        );
      }
    } else if (event.event === 'subscription.cancelled' || event.event === 'subscription.completed') {
      const notes = event.payload?.subscription?.entity?.notes || {};
      const { org_id } = notes;
      if (org_id) {
        await pool.query(
          `UPDATE organizations SET plan = 'free', subscription_status = 'cancelled', subscription_id = NULL WHERE id = $1`,
          [org_id]
        );
      }
    }
    return res.json({ received: true });
  } catch (err) {
    logger.error(`Webhook processing error: ${err.message}`);
    return res.status(500).json({ detail: 'Webhook processing failed.' });
  }
});

export default router;
